src/main/kotlin/io/libp2p/core/Discoverer.kt

```diff
@@ -0,0 +1,11 @@
+package io.libp2p.core
+
+import java.util.concurrent.CompletableFuture
+
+typealias PeerListener = (PeerInfo) -> Unit
+
+interface Discoverer {
+    fun start(): CompletableFuture<Void>
+    fun stop(): CompletableFuture<Void>
+    fun onPeerFound(listener: PeerListener)
```

Comment:
From my perspective it's a bit odd name for the method for adding a listener. 
For me the following looks nice: 
```kt
interface Discoverer {
    fun start(): CompletableFuture<Void>
    fun stop(): CompletableFuture<Void>
    val newPeerFoundListeners: MutableCollection<PeerListener>
}
...
discoverer.newPeerFoundListeners += { println(it) }
```
However this is a matter of taste